Optimize `panic!(<str lit>)` into method call

jhpratt opened the pull request:
In some places, the standard library calls out to a standalone method (`core::panicking::panic`) when panicking to avoid unnecessary use of the formatting machinery. This PR attempts to optimize `panic!` calls directly and for everyone, rather than requiring manual effort to call the method.

The current approach is to intercept `panic!` and `unreachable!` when it's going thru edition resolution. If any only if the macro call consists of exactly one string literal _and_ there are no formatting placeholders (determined laxly by searching for braces in the string), then we pass the message into the method call. In the case of `unreachable!`, we need to prepend the message prefix, necessitating unescaping as that info isn't saved at this point of the compilation process.

As with any optimization, I _think_ this will work, but a perf run will be necessary.

jhpratt committed:
Optimize `panic!(<str lit>)` into method call

jhpratt committed:
Update `incompatible_msrv` to ignore all of stdlib

This ignores `core`, `alloc`, and `std` rather than just `core`. The
same reasoning applies.

jhpratt commented:
@bors2 try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

jhpratt commented:
Awesome, it actually worked!

Marking as ready for review, though I'm going to add some doc changes and misc. cleanups in nearby code. The existing changes won't be altered.

jhpratt commented:
re. the pings: miri is simply blessing a test since the backtrace changed (expected), and clippy is updating the MSRV lint to handle `core`, `alloc`, and `std` instead of just `core`. This is necessary as a test would fail without `std` being included.

jhpratt committed:
Use helper methods when possible

lcnr reviewed (commented):
am not too familiar with this code, but it looks good to me. cc @rust-lang/libs on whether this is generally something you want

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/148375/b15a874aafe7eab9ea3ac2c1d59c7b03e1425027..e890c3b2f07707255a51d699b4f49cef8c8ae9b4)*

Urgau commented on compiler/rustc_builtin_macros/src/edition_panic.rs:
```diff
@@ -19,50 +40,99 @@ pub(crate) fn expand_panic<'cx>(
+        && let TokenKind::Literal(lit) = &token.kind
+        && let Lit { kind: LitKind::Str | LitKind::StrRaw(_), symbol, .. } = lit
+        && let msg = symbol.as_str()
+        && !msg.contains(|c| c == '{' || c == '}')
```
This is very much a hack. It would be much better to use `rustc_parse_format` here.

In particular since I think your version would not optimize `panic!("foo {{ bar }}");` since it doesn't know about the escaping.

jhpratt commented:
> My first impression of this PR is that it seems like a lot of complexity on the macro expansion side, and that it feels like this should be doable by using `fmt::Arguments::as_str()` (or `as_statically_known_str()`) instead. Have you tried anything like that? (E.g. an inline(always) function with an `if let Some(m) = m.as_statically_known_str()` dispatching to either panic() or panic_fmt()?)

It didn't cross my mind. I'll push up a commit with just that and kick off a perf run to see the effect, both relative and absolute.

(@lcnr: removing your assignment due to your comment and Mara's self-assignment)

jhpratt committed:
Optimize in macro call directly

This excludes `panic_2015!` as calling it in the same manner results in
triggering the edition lint in a few places.

m-ou-se commented:
> A more extreme approach would be to optimise `fmt::Arguments` such that it's the same size as a `&str` and have a trivial conversion from `&str` to `fmt::Arguments`. Then `panic()` has no more advantage over `panic_fmt()`. I've been wanting to do that for years, but kept running into roadblocks. But I think many of them have been resolved; it is probably easier now. I'll investigate.

Trying out some stuff in https://github.com/rust-lang/rust/pull/148529.
